specification/sections/03-evaluation-context.md

```diff
@@ -42,15 +42,39 @@ see: [structure](../types.md#structure), [datetime](../types.md#datetime)
 
 The key uniquely identifies a field in the `evaluation context` and it should be unique across all types to avoid any collision when marshalling the `evaluation context` by the provider.
 
-### 3.2 Merging Context
+### 3.2 Context levels and merging
 
-#### Requirement 3.2.1
+#### Condition 3.2.1
+
+> The implementation uses the single-context paradigm.
+
+see: [single-context paradigm](../glossary.md#single-context-paradigm)
+
+##### Conditional Requirement 3.2.1.1
+
+> The API **MUST** have a method for supplying `evaluation context`.
+
+API (global) `evaluation context` can be used to supply data to flag evaluation, such as (but not limited to) user name, email, or shopping cart items.
+
+##### Conditional Requirement 3.2.1.2
+
+> The Client and invocation **MUST NOT** have a method for supplying `evaluation context`.
+
+In the single-context paradigm, context is global. The client and invocation cannot supply evaluation context. 
```

Comment:
I'm struggling with this quite a bit. 

On the one hand, I'm concerned that making evaluation context global is going to cause serious issue for sophisticated consumers who target flag evaluation across multiple dimensions. For example, the Spotify use case where some flags vary by artists, and the "current artist" is going to change frequently. If we only have 1 single shared context then every time ANY evaluation context changes then ALL feature flags would need to be re-evaluated. If we had some sort of nesting/merging capability then we could probably avoid this "cache busting" problem.

However, on the other hand, I suspect that some provider's client-side SDKs assume that evaluation context is shared global state, and if we allowed nested state that could make implementing the OF provider API harder.

Reply:
> On the one hand, I'm concerned that making evaluation context global is going to cause serious issue for sophisticated consumers who target flag evaluation across multiple dimensions. For example, the Spotify use case where some flags vary by artists, and the "current artist" is going to change frequently. If we only have 1 single shared context then every time ANY evaluation context changes then ALL feature flags would need to be re-evaluated. If we had some sort of nesting/merging capability then we could probably avoid this "cache busting" problem.

Especially with the ability to dif the old/new context, could this be something the provider handles? For example, my context could be a complex object with multiple top level objects, and I could check them for changes in `onContextSet` and only re-evaluate what's needed, assuming my provider is intelligent enough for that.